core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java

```diff
@@ -305,25 +316,46 @@ private boolean notifySubscriber(Subscriber<Object> subscriber, Queue<Object> qu
 
             if (demand == Long.MAX_VALUE || demandUpdater.compareAndSet(this, demand, demand - 1)) {
                 @SuppressWarnings("unchecked")
-                final T o = (T) queue.remove();
+                T o = (T) queue.remove();
+                ReferenceCountUtil.touch(o);
                 onRemoval(o);
-                if (!subscription.withPooledObjects() && o instanceof ByteBufHttpData) {
-                    ByteBuf buf = ((ByteBufHttpData) o).buf();
-                    try {
-                        subscriber.onNext(HttpData.of(ByteBufUtil.getBytes(buf)));
-                    } finally {
-                        buf.release();
+                if (!subscription.withPooledObjects()) {
+                    if (o instanceof ByteBufHolder) {
```

Comment:
This is just a code clean-up which is not related with a leak. I just handled ByteBuf additionally because StreamMessage is rather a generic type which can handle any objects as well as HttpObject. So, your understanding is correct. :-)